server/storage/mvcc/revision.go

```diff
@@ -14,48 +14,89 @@
 
 package mvcc
 
-import "encoding/binary"
+import (
+	"encoding/binary"
+)
 
-// revBytesLen is the byte length of a normal revision.
-// First 8 bytes is the revision.main in big-endian format. The 9th byte
-// is a '_'. The last 8 bytes is the revision.sub in big-endian format.
-const revBytesLen = 8 + 1 + 8
+const (
+	// revBytesLen is the byte length of a normal revision.
+	// First 8 bytes is the revision.main in big-endian format. The 9th byte
+	// is a '_'. The last 8 bytes is the revision.sub in big-endian format.
+	revBytesLen = 8 + 1 + 8
+	// markedRevBytesLen is the byte length of marked revision.
+	// The first `revBytesLen` bytes represents a normal revision. The last
+	// one byte is the mark.
+	markedRevBytesLen      = revBytesLen + 1
+	markBytePosition       = markedRevBytesLen - 1
+	markTombstone     byte = 't'
+)
 
-// A revision indicates modification of the key-value space.
-// The set of changes that share same main revision changes the key-value space atomically.
-type revision struct {
-	// main is the main revision of a set of changes that happen atomically.
-	main int64
-
-	// sub is the sub revision of a change in a set of changes that happen
+type Revision struct {
+	// Main is the main revision of a set of changes that happen atomically.
+	Main int64
+	// Sub is the sub revision of a change in a set of changes that happen
 	// atomically. Each change has different increasing sub revision in that
 	// set.
-	sub int64
+	Sub int64
 }
 
-func (a revision) GreaterThan(b revision) bool {
-	if a.main > b.main {
+func (a Revision) GreaterThan(b Revision) bool {
+	if a.Main > b.Main {
 		return true
 	}
-	if a.main < b.main {
+	if a.Main < b.Main {
 		return false
 	}
-	return a.sub > b.sub
+	return a.Sub > b.Sub
+}
+
+// BucketKey indicates modification of the key-value space.
+// The set of changes that share same main revision changes the key-value space atomically.
+type BucketKey struct {
+	Revision
+	tombstone bool
 }
 
-func newRevBytes() []byte {
+func newBucketKey(main, sub int64, isTombstone bool) BucketKey {
+	return BucketKey{
+		Revision: Revision{
+			Main: main,
+			Sub:  sub,
+		},
+		tombstone: isTombstone,
+	}
+}
+
+func NewRevBytes() []byte {
 	return make([]byte, revBytesLen, markedRevBytesLen)
 }
 
-func revToBytes(rev revision, bytes []byte) {
-	binary.BigEndian.PutUint64(bytes, uint64(rev.main))
+func RevToBytes(rev BucketKey, bytes []byte) []byte {
```

Comment:
```suggestion
func BucketKeyToBytes(rev BucketKey, bytes []byte) []byte {
```

Also add a separate `RevToBytes` something like below,
```
func RevToBytes(rev Revision, bytes []byte) []byte {
    return BucketKeyToBytes(newBucketKey(rev.Main, rev.Sub, false))
}
```